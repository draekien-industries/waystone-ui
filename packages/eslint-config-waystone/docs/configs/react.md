# eslint-config-waystone - react

## Usage

```js
// .eslintrc.js
module.exports = {
  extends: ['waystone/react'],
};
```

## Env

- browser
- es6
- shared-node-browser

## Extends

- airbnb
- turbo
- eslint:recommended
- import/recommended
- import/react
- import/typescript
- @typescript-eslint/eslint-recommended
- @typescript-eslint/recommended
- react/recommended
- react/jsx-runtime
- waystone/recommended
- jsx-a11y/recomended
- testing-library/recommended
- jest/recommended
- prettier
