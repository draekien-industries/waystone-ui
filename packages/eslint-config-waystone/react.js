module.exports = {
  env: {
    browser: true,
    es6: true,
    'shared-node-browser': true,
  },
  extends: [
    'airbnb',
    'turbo',
    'eslint:recommended',
    'plugin:import/recommended',
    'plugin:import/react',
    'plugin:import/typescript',
    'plugin:@typescript-eslint/eslint-recommended',
    'plugin:@typescript-eslint/recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:react-hooks/recommended',
    'plugin:waystone/recommended',
    'plugin:jsx-a11y/recommended',
    'prettier',
  ],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: {
      jsx: true,
    },
    ecmaVersion: 'latest',
    sourceType: 'module',
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
  rules: {
    'no-plusplus': ['error', { allowForLoopAfterthoughts: true }],
    '@typescript-eslint/consistent-type-definitions': ['error', 'type'],
    '@typescript-eslint/consistent-type-imports': [
      'error',
      {
        prefer: 'type-imports',
        fixStyle: 'separate-type-imports',
      },
    ],
    '@typescript-eslint/no-unused-vars': [
      'error',
      { varsIgnorePattern: '^_*' },
    ],
    'react/function-component-definition': [
      'warn',
      {
        namedComponents: 'arrow-function',
        unnamedComponents: 'arrow-function',
      },
    ],
    'react/jsx-filename-extension': ['warn', { extensions: ['.tsx'] }],
    'react/jsx-props-no-spreading': 'off',
    'react/no-unknown-property': ['error', { ignore: ['sx'] }],
    'react/require-default-props': 'off',
    'import/extensions': [
      'error',
      'ignorePackages',
      { ts: 'never', tsx: 'never', js: 'never', jsx: 'never' },
    ],
    'import/prefer-default-export': 'off',
    'import/no-extraneous-dependencies': ['error', { devDependencies: true }],
    'no-use-before-define': [
      'error',
      {
        functions: false,
        classes: true,
        variables: true,
        allowNamedExports: false,
      },
    ],
  },
  overrides: [
    {
      env: { jest: true },
      files: [
        '**/__tests__/**/*.[jt]s?(x)',
        '**/__fixtures__/**/*.[jt]s?(x)',
        '**/?(*.)+(spec|test).[jt]s?(x)',
      ],
      extends: [
        'airbnb',
        'turbo',
        'eslint:recommended',
        'plugin:import/recommended',
        'plugin:import/typescript',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:react/recommended',
        'plugin:react/jsx-runtime',
        'plugin:react-hooks/recommended',
        'plugin:waystone/recommended',
        'plugin:jsx-a11y/recommended',
        'plugin:testing-library/react',
        'plugin:jest/recommended',
        'prettier',
      ],
      rules: {
        'import/no-extraneous-dependencies': [
          'off',
          { devDependencies: ['**/?(*.)+(spec|test).[jt]s?(x)'] },
        ],
        'react/jsx-filename-extension': ['warn', { extensions: ['.tsx'] }],
        'import/extensions': [
          'error',
          'ignorePackages',
          { ts: 'never', tsx: 'never', js: 'never', jsx: 'never' },
        ],
      },
    },
  ],
  ignorePatterns: [
    '**/*.js',
    '**/*.json',
    '**/*.d.ts',
    '**/*.g.*',
    '**/*.generated.*',
    'node_modules',
    'public',
    'styles',
    '.next',
    'coverage',
    'dist',
    '.turbo',
  ],
};
