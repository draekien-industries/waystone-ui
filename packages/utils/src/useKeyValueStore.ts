import { useCallback, useMemo, useReducer, type Reducer } from 'react';

/**
 * An alias for a record containing unknown keys and values
 */
export type KeyValueStore = Record<PropertyKey, unknown>;

/**
 * Extracts the value types associated with the keys of a key value store into a tuple
 * @template T a record of type KeyValueStore
 * @template K the keys being used to access the KeyValueStore
 * @template Acc the accumulator holding the constructed tuple
 * @see {@link KeyValueStore}
 */
export type ValuesFromKeyValueStore<
  T extends KeyValueStore,
  K extends Array<keyof T>,
  Acc extends Array<unknown> = [],
> = K extends [infer First, ...infer Rest]
  ? First extends keyof T
    ? Rest extends Array<keyof T>
      ? ValuesFromKeyValueStore<T, Rest, [...Acc, T[First] | null]>
      : Acc
    : Acc
  : Acc;

type KeyValueStoreAction<T extends KeyValueStore> =
  | {
      type: 'SET_VALUE';
      key: keyof T;
      value: unknown;
    }
  | {
      type: 'SET_VALUES';
      values: Partial<T>;
    }
  | {
      type: 'CLEAR_VALUES';
      keys: Array<keyof T>;
    };

const keyValueStoreReducer = <T extends KeyValueStore>(
  state: T,
  action: KeyValueStoreAction<T>
) => {
  switch (action.type) {
    case 'SET_VALUE':
      return { ...state, [action.key]: action.value };
    case 'SET_VALUES':
      return { ...state, ...action.values };
    case 'CLEAR_VALUES':
      return action.keys.reduce(
        (acc, key) => ({
          ...acc,
          [key]: null,
        }),
        state
      );
    default:
      return state;
  }
};

/**
 * The entire {@link KeyValueStore} if no keys are provided, otherwise the {@link ValuesFromKeyValueStore}
 */
export type KeyValueStoreGetValuesReturn<
  T extends KeyValueStore,
  K extends Array<keyof T>,
> = K extends [] ? T : ValuesFromKeyValueStore<T, K>;

/**
 * A hook which defines methods for interacting with some object saved in react state.
 * @template T the type of the KeyValueStore.
 * @param initialStore an optional initial value to assign as the {@link KeyValueStore}. Defaults to `{}` when not provided.
 * @returns methods for interacting with the key value store. See {@link UseKeyValueStoreReturn} for more details.
 */
export const useKeyValueStore = <T extends KeyValueStore>(initialStore?: T) => {
  const [state, dispatch] = useReducer<Reducer<T, KeyValueStoreAction<T>>>(
    keyValueStoreReducer,
    initialStore ?? ({} as T)
  );

  /**
   * Gets values from the key value store
   * @param keys an optional array of keys that should be used to access the key value store
   * @returns The entire {@link KeyValueStore} if no keys are provided, otherwise the {@link ValuesFromKeyValueStore}
   */
  const getValues = useCallback(
    <K extends Array<keyof T>>(
      ...keys: K
    ): KeyValueStoreGetValuesReturn<T, K> => {
      if (keys.length) {
        return keys.map((key) => state[key]) as KeyValueStoreGetValuesReturn<
          T,
          K
        >;
      }

      return state as KeyValueStoreGetValuesReturn<T, K>;
    },
    [state]
  );

  /**
   * Updates the key value store, setting the value associated with the provided key to the new value
   * @param key a key present in the key value store
   * @param value the value to assign the specified key
   */
  const setValue = useCallback((key: keyof T, value: T[typeof key]) => {
    dispatch({
      type: 'SET_VALUE',
      key,
      value,
    });
  }, []);

  /**
   * Updates the key value store with the provided values, replacing any matching keys with their updated values via a shallow merge.
   * @param values the values that should be updated in the store
   */
  const setValues = useCallback((values: Partial<T>) => {
    dispatch({
      type: 'SET_VALUES',
      values,
    });
  }, []);

  /**
   * Updates the key value store, setting the values of the provided keys to `null`
   */
  const clearValues = useCallback((...keys: Array<keyof T>) => {
    dispatch({ type: 'CLEAR_VALUES', keys });
  }, []);

  return useMemo(
    () => ({
      clearValues,
      getValues,
      setValue,
      setValues,
    }),
    [clearValues, getValues, setValue, setValues]
  );
};

export type UseKeyValueStoreReturn<T extends KeyValueStore = KeyValueStore> =
  ReturnType<typeof useKeyValueStore<T>>;
