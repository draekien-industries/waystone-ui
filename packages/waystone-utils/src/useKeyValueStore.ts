/* eslint-disable @typescript-eslint/no-explicit-any */
import { useCallback, useState } from 'react';

export type GetValueOfKeyParams<
  TStore extends Record<string, any>,
  TKey extends keyof TStore
> = {
  key: TKey;
};

export type GetValueOfKeyReturn<
  TStore extends Record<string, any>,
  TKey extends keyof TStore
> = TStore[TKey];

export type SetValueOfKeyParams<
  TStore extends Record<string, any>,
  TKey extends keyof TStore
> = {
  key: TKey;
  value: TStore extends Record<string, infer TValue> ? TValue : never;
};

export const useStrictKeyValueStore = <
  TStore extends Record<string, any> = Record<string, any>
>(
  initialStore = {} as TStore
) => {
  const [store, setStore] = useState<TStore>(initialStore);

  const getValueOfKey = useCallback(
    <TKey extends keyof TStore>({
      key,
    }: GetValueOfKeyParams<TStore, TKey>): GetValueOfKeyReturn<TStore, TKey> =>
      store[key],
    [store]
  );

  const setValueOfKey = useCallback(
    ({ key, value }: SetValueOfKeyParams<TStore, keyof TStore>) => {
      setStore((prev) => {
        const copy = { ...prev };

        copy[key] = value;

        return copy;
      });
    },
    [setStore]
  );

  return {
    store,
    getValueOfKey,
    setValueOfKey,
  };
};

export type UseStrictKeyValueStoreReturn<
  TStore extends Record<string, any> = Record<string, any>
> = ReturnType<typeof useStrictKeyValueStore<TStore>>;

export const useKeyValueStore = (initialStore = {} as Record<string, any>) => {
  const [store, setStore] = useState(initialStore);

  const getValueOfKey = useCallback(
    <TValue = unknown>({
      key,
    }: GetValueOfKeyParams<Record<string, any>, string>): TValue => {
      if (key in store) {
        const value = store[key];

        return value;
      }

      throw new Error(
        `The key ${key} does not exist in the store. Current keys are: [${Object.keys(
          store
        ).join(',')}]`
      );
    },
    [store]
  );

  const setValueOfKey = useCallback(
    ({ key, value }: SetValueOfKeyParams<Record<string, any>, string>) => {
      setStore((prev) => {
        const copy = { ...prev };

        copy[key] = value;

        return copy;
      });
    },
    [setStore]
  );

  return {
    store,
    getValueOfKey,
    setValueOfKey,
  };
};

export type UseKeyValueStoreReturn = ReturnType<typeof useKeyValueStore>;
