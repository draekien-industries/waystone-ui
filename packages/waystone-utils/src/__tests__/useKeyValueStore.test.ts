import { act, renderHook } from '@testing-library/react';
import { useKeyValueStore, useStrictKeyValueStore } from '../useKeyValueStore';

describe('useStrictKeyValueStore hook', () => {
  it('returns the default values when no updates are triggered', () => {
    const { result } = renderHook(() =>
      useStrictKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    expect(result.current.getValueOfKey({ key: '0' })).toStrictEqual({
      test: 'one',
    });

    expect(result.current.getValueOfKey({ key: '1' })).toStrictEqual({
      test: 'two',
    });

    expect(result.current.store).toStrictEqual({
      '0': { test: 'one' },
      '1': { test: 'two' },
    });
  });

  it('returns the new value of a given key after it has been set', () => {
    const { result } = renderHook(() =>
      useStrictKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    act(() => {
      result.current.setValueOfKey({ key: '0', value: { test: 'new one' } });
    });

    expect(result.current.getValueOfKey({ key: '0' })).toStrictEqual({
      test: 'new one',
    });
  });

  it('throws an error when trying to access a key that does not exist', () => {
    const { result } = renderHook(() =>
      useStrictKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    act(() => {
      // @ts-expect-error key does not exist
      result.current.setValueOfKey({ key: '2', value: { test: 'new one' } });
    });

    // @ts-expect-error key does not exist
    expect(result.current.getValueOfKey({ key: '2' })).toStrictEqual({
      test: 'new one',
    });
  });
});

describe('useKeyValueStore hook', () => {
  it('returns the default values when no updates are triggered', () => {
    const { result } = renderHook(() =>
      useKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    expect(result.current.getValueOfKey({ key: '0' })).toStrictEqual({
      test: 'one',
    });

    expect(result.current.getValueOfKey({ key: '1' })).toStrictEqual({
      test: 'two',
    });

    expect(result.current.store).toStrictEqual({
      '0': { test: 'one' },
      '1': { test: 'two' },
    });
  });

  it('returns the new value of a given key after it has been set', () => {
    const { result } = renderHook(() =>
      useKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    act(() => {
      result.current.setValueOfKey({ key: '0', value: { test: 'new one' } });
    });

    expect(result.current.getValueOfKey({ key: '0' })).toStrictEqual({
      test: 'new one',
    });
  });

  it('can set the value of a key that does not exist initially', () => {
    const { result } = renderHook(() =>
      useKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    act(() => {
      result.current.setValueOfKey({ key: '2', value: { test: 'new one' } });
    });

    expect(result.current.getValueOfKey({ key: '2' })).toStrictEqual({
      test: 'new one',
    });
  });

  it('throws an error when trying to access a key that does not exist', () => {
    const { result } = renderHook(() =>
      useKeyValueStore({
        '0': { test: 'one' },
        '1': { test: 'two' },
      })
    );

    expect(() =>
      result.current.getValueOfKey({ key: '2' })
    ).toThrowErrorMatchingSnapshot();
  });
});
