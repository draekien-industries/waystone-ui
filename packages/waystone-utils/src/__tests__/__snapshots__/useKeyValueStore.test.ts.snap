// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useKeyValueStore hook throws an error when trying to access a key that does not exist 1`] = `"The key 2 does not exist in the store. Current keys are: [0,1]"`;
